Source/WebKit/NetworkProcess/NetworkDataTaskDataURL.cpp

```diff
@@ -0,0 +1,199 @@
+#include "NetworkSession.h"
+#include <WebCore/ResourceError.h>
+#include <WebCore/ResourceRequest.h>
+#include <WebCore/ResourceResponse.h>
+#include <pal/text/TextEncoding.h>
+#include <wtf/RunLoop.h>
+
+#if USE(CURL)
+#include <curl/curl.h>
+#elif USE(SOUP)
+#include "WebErrors.h"
+#endif
+
+namespace WebKit {
+using namespace WebCore;
+
+Ref<NetworkDataTask> NetworkDataTaskDataURL::create(NetworkSession& session, NetworkDataTaskClient& client, const NetworkLoadParameters& parameters)
+{
+    ASSERT(parameters.request.url().protocolIsData());
+    return adoptRef(*new NetworkDataTaskDataURL(session, client, parameters));
+}
+
+NetworkDataTaskDataURL::NetworkDataTaskDataURL(NetworkSession& session, NetworkDataTaskClient& client, const NetworkLoadParameters& parameters)
+    : NetworkDataTask(session, client, parameters.request, parameters.storedCredentialsPolicy, parameters.shouldClearReferrerOnHTTPSToHTTPRedirect, parameters.isMainFrameNavigation)
+{
+    m_session->registerNetworkDataTask(*this);
+}
+
+NetworkDataTaskDataURL::~NetworkDataTaskDataURL()
+{
+    invalidateAndCancel();
+}
+
+void NetworkDataTaskDataURL::resume()
+{
+    ASSERT(m_state != State::Running);
+    if (m_state == State::Canceling || m_state == State::Completed)
+        return;
+
+    m_state = State::Running;
+
+    DataURLDecoder::decode(firstRequest().url(), { }, [this, protectedThis = Ref { *this }](auto decodeResult) mutable {
+        if (m_state == State::Canceling || m_state == State::Completed)
+            return;
+
+        m_dataURLResult = WTFMove(decodeResult);
+        didDecodeDataURL();
+    });
+}
+
+void NetworkDataTaskDataURL::cancel()
+{
+    if (m_state == State::Canceling || m_state == State::Completed)
+        return;
+
+    m_state = State::Canceling;
+}
+
+void NetworkDataTaskDataURL::invalidateAndCancel()
+{
+    cancel();
+    m_state = State::Completed;
+}
+
+NetworkDataTask::State NetworkDataTaskDataURL::state() const
+{
+    return m_state;
+}
+
+void NetworkDataTaskDataURL::setPendingDownloadLocation(const String& filename, SandboxExtension::Handle&& sandboxExtensionHandle, bool allowOverwrite)
+{
+    NetworkDataTask::setPendingDownloadLocation(filename, WTFMove(sandboxExtensionHandle), allowOverwrite);
+    m_allowOverwriteDownload = allowOverwrite;
+}
+
+String NetworkDataTaskDataURL::suggestedFilename() const
+{
+    if (!m_suggestedFilename.isEmpty())
+        return m_suggestedFilename;
+
+    String suggestedFilename = m_response.suggestedFilename();
+    if (!suggestedFilename.isEmpty())
+        return suggestedFilename;
+
+    return PAL::decodeURLEscapeSequences(m_response.url().lastPathComponent());
+}
+
+void NetworkDataTaskDataURL::didDecodeDataURL()
+{
+    ASSERT(m_state == State::Running);
+    ASSERT(m_decodingState == DecodingState::Finished);
+    if (!m_dataURLResult) {
+        if (m_client)
+            m_client->didCompleteWithError(internalError(firstRequest().url()));
+        return;
```

Comment:
Please change m_state to State::Completed after error occurs. This will prevent malfunctions.

Reply:
You should call invalidateAndCancel() here.